File: Code/Aula_15/Ex3.c
#include <stdio.h>
#include <fcntl.h>
#include <sys/poll.h>
#include <unistd.h>

int main(void)
{
	struct pollfd pfd;
	char buffer;
	int btn_press;
	system("echo 4       > /sys/class/gpio/export");
	system("echo falling > /sys/class/gpio/gpio4/edge");
	system("echo in      > /sys/class/gpio/gpio4/direction");
	pfd.fd = open("/sys/class/gpio/gpio4/value", O_RDONLY);
	if(pfd.fd < 0)
	{
		puts("Erro abrindo /sys/class/gpio/gpio4/value");
		puts("Execute este programa como root.");
		return 1;
	}
	pfd.events = POLLPRI | POLLERR;
	pfd.revents = 0;
	puts("Pressione o botao...");
	for(btn_press=0; btn_press<5; btn_press++)
	{
		read(pfd.fd, &buffer, 1);
		poll(&pfd, 1, -1);
		puts("Borda de descida!");
	}
	close(pfd.fd);
	system("echo 4 > /sys/class/gpio/unexport");
	return 0;
}

File: 3_Programação_RPi/3.3_GPIO_Polling/Python/Ex3.py
import RPi.GPIO as gpio
import os, select

pin = 4
os.system("echo %d      > /sys/class/gpio/export" % pin)
os.system("echo falling > /sys/class/gpio/gpio%d/edge" % pin)
os.system("echo in      > /sys/class/gpio/gpio%d/direction" % pin)

fd = os.open("/sys/class/gpio/gpio%d/value" % pin, os.O_RDONLY)
pfd = select.poll()
pfd.register(fd, select.POLLPRI | select.POLLERR)

print("Pressione o botao...")
for btn_press in range(5):
	os.lseek(fd, 0, os.SEEK_SET)
	os.read(fd, 1)
	pfd.poll()
	print("Borda de descida!")

os.close(fd)
os.system("echo %d > /sys/class/gpio/unexport" % pin)
